{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.8.0'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import HTSeq\n",
    "HTSeq.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sam_fpath='/ifs/home/yy1533/Lab/cel-seq-pipe/demo/celseq2/smallsam/BC-15-TCTGAG.sam'\n",
    "gff_fpath='/ifs/data/yanailab/refs/danio_rerio/danRer10_87/gtf/Danio_rerio.GRCz10.87.gtf.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "htseq_fpath='/ifs/home/yy1533/Lab/cel-seq-pipe/demo/celseq2/sam2umicnt/BC-15-TCTGAG.count.bk' ## generated by default htseq-count command"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MWE for single-end sam + GFF => alignments count\n",
    "\n",
    "<http://htseq.readthedocs.io/en/master/counting.html>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Prepare GFF\n",
    "fh_gff = HTSeq.GFF_Reader(gff_fpath)\n",
    "features = HTSeq.GenomicArrayOfSets( \"auto\", stranded=True )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "feature_atrr = 'gene_id'\n",
    "feature_type = 'exon'\n",
    "accept_aln_qual_min = 10\n",
    "is_gapped_aligner = True ## bowtie2 is False\n",
    "## Bowtie2: 1) non_gapped_aligner; 2) default report best one for multiple alignment\n",
    "len_umi = 6\n",
    "len_bc = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for gff in fh_gff:\n",
    "    if gff.type != feature_type:\n",
    "        continue\n",
    "    features[gff.iv] += gff.attr[feature_atrr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "# pickle.dump(features, open('test.p', 'wb'))\n",
    "\n",
    "# %%time\n",
    "# features_bk = pickle.load(open('test.p', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## sam per cell\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = Counter()\n",
    "fh_aln = HTSeq.SAM_Reader(sam_fpath)\n",
    "for aln in fh_aln:\n",
    "    if not aln.aligned:\n",
    "        counts[\"_unmapped\"] += 1\n",
    "        continue\n",
    "    if aln.aQual < accept_aln_qual_min:\n",
    "        counts[\"_low_qual\"] += 1\n",
    "        continue\n",
    "#     try: # bowtie2 report best one randomly by default\n",
    "#         if aln.optional_field( \"NH\" ) > 1:\n",
    "#             counts['_multimapped'] += 1\n",
    "#             continue\n",
    "#     except KeyError:\n",
    "#         pass\n",
    "    \n",
    "    gene_ids = set()\n",
    "    if is_gapped_aligner:\n",
    "        for aln_part in aln.cigar:\n",
    "            if aln_part.type != 'M':\n",
    "                continue\n",
    "            for _, gene_id in features[aln_part.ref_iv].steps():\n",
    "                gene_ids |= gene_id\n",
    "    else: # bowtie2 is non-gapped-aligner\n",
    "        for _, gene_id in features[aln.iv].steps():\n",
    "            gene_ids |= gene_id\n",
    "\n",
    "    ## union model        \n",
    "    if len(gene_ids) == 1:\n",
    "        gene_id = list(gene_ids)[0]\n",
    "        counts[gene_id] += 1\n",
    "    elif len(gene_ids) == 0:\n",
    "        counts[\"_no_feature\"] += 1\n",
    "    else:\n",
    "        counts[\"_ambiguous\"] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSDARG00000041450 1\n",
      "_unmapped 1562\n",
      "ENSDARG00000014690 115\n",
      "_low_qual 2023\n",
      "ENSDARG00000080337 339\n",
      "ENSDARG00000036162 7\n",
      "ENSDARG00000018334 4\n",
      "_no_feature 2100\n",
      "ENSDARG00000103791 3\n",
      "ENSDARG00000032430 1\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "for gene_id in counts:\n",
    "    print(gene_id, counts[gene_id])\n",
    "    i += 1\n",
    "    if i == 10: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;31m\u001b[KENSDARG00000080337\u001b[m\u001b[K\t339\n",
      "\u001b[01;31m\u001b[KENSDARG00000014690\u001b[m\u001b[K\t115\n"
     ]
    }
   ],
   "source": [
    "! grep ENSDARG00000080337 {htseq_fpath}\n",
    "! grep ENSDARG00000014690 {htseq_fpath}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MWE for sam+GFF => UMI count "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fh_aln = HTSeq.SAM_Reader(sam_fpath)\n",
    "\n",
    "aln = next(iter(fh_aln))\n",
    "\n",
    "aln.read.name\n",
    "\n",
    "aln.read.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CGTTAC\n",
      "CGTTAC\n"
     ]
    }
   ],
   "source": [
    "def _umi_seq(name, length=6):\n",
    "    ## BC-TCTGAG_UMI-CGTTAC => \n",
    "    try:\n",
    "        out = name.split('_')[1][4:4+length]\n",
    "    except Exception as e:\n",
    "        raise(e)\n",
    "    return(out)\n",
    "foox='BC-TCTGAG_UMI-CGTTAC'\n",
    "foo=_umi_seq(foox, 6)\n",
    "print(foo)\n",
    "foox='BC-TCTGAG_UMI-AAAAAA'\n",
    "print(foo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "umi_cnt = defaultdict(set)\n",
    "aln_cnt = Counter()\n",
    "fh_aln = HTSeq.SAM_Reader(sam_fpath)\n",
    "i = 0\n",
    "for aln in fh_aln:\n",
    "    i += 1\n",
    "    if not aln.aligned:\n",
    "        aln_cnt[\"_unmapped\"] += 1\n",
    "        continue\n",
    "    if aln.aQual < accept_aln_qual_min:\n",
    "        aln_cnt[\"_low_qual\"] += 1\n",
    "        continue\n",
    "    try: # bowtie2 report best one randomly by default\n",
    "        if aln.optional_field( \"NH\" ) > 1:\n",
    "            aln_cnt['_multimapped'] += 1\n",
    "            continue\n",
    "    except KeyError:\n",
    "        pass\n",
    "    \n",
    "    gene_ids = set()\n",
    "    if is_gapped_aligner:\n",
    "        for aln_part in aln.cigar:\n",
    "            if aln_part.type != 'M':\n",
    "                continue\n",
    "            for _, gene_id in features[aln_part.ref_iv].steps():\n",
    "                gene_ids |= gene_id\n",
    "    else: # bowtie2 is non-gapped-aligner\n",
    "        for _, gene_id in features[aln.iv].steps():\n",
    "            gene_ids |= gene_id\n",
    "    ## union model        \n",
    "    if len(gene_ids) == 1:\n",
    "        gene_id = list(gene_ids)[0]\n",
    "        aln_cnt[gene_id] += 1\n",
    "        umi_seq = _umi_seq(aln.read.name)\n",
    "        umi_cnt[gene_id].add(umi_seq)\n",
    "    elif len(gene_ids) == 0:\n",
    "        aln_cnt[\"_no_feature\"] += 1\n",
    "    else:\n",
    "        aln_cnt[\"_ambiguous\"] += 1        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "umi_count = Counter({x : len(umi_cnt.get(x, set())) for x in umi_cnt})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSDARG00000041450 1\n",
      "ENSDARG00000014690 107\n",
      "ENSDARG00000080337 285\n",
      "ENSDARG00000036162 7\n",
      "ENSDARG00000018334 4\n",
      "ENSDARG00000103791 3\n",
      "ENSDARG00000032430 1\n",
      "ENSDARG00000051888 8\n",
      "ENSDARG00000051975 1\n",
      "ENSDARG00000024540 9\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "for gene_id in umi_count:\n",
    "    print(gene_id, umi_count[gene_id])\n",
    "    i += 1\n",
    "    if i == 10: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;31m\u001b[KENSDARG00000014690\u001b[m\u001b[K\t115\r\n"
     ]
    }
   ],
   "source": [
    "! grep ENSDARG00000014690 {htseq_fpath}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Test modules\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[ Wed Jun 28 11:47:02 2017 ] Processing 0 lines of GFF...\n",
      "[ Wed Jun 28 11:47:41 2017 ] Processing 100,000 lines of GFF...\n",
      "[ Wed Jun 28 11:48:20 2017 ] Processing 200,000 lines of GFF...\n",
      "[ Wed Jun 28 11:49:00 2017 ] Processing 300,000 lines of GFF...\n",
      "[ Wed Jun 28 11:49:39 2017 ] Processing 400,000 lines of GFF...\n"
     ]
    }
   ],
   "source": [
    "from celseq2.prepare_annotation_model import cook_anno_model\n",
    "features2 = cook_anno_model(gff_fpath, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "from celseq2.count_umi import count_umi\n",
    "features2 = pickle.load(open('./test.p', 'rb'))\n",
    "umi_count2 = count_umi('/ifs/home/yy1533/Lab/cel-seq-pipe/demo/celseq2/smallsam/BC-15-TCTGAG.sam',\n",
    "                       features2, 6, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umi_count2['ENSDARG00000014690']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
